Source/Core/Core/IOS/USB/Emulated/Skylanders/SkylanderFigure.cpp

```diff
@@ -0,0 +1,455 @@
+  // the start of the string is intentional
+  static constexpr std::array<u8, 0x35> HASH_CONST = {
+      0x20, 0x43, 0x6F, 0x70, 0x79, 0x72, 0x69, 0x67, 0x68, 0x74, 0x20, 0x28, 0x43, 0x29,
+      0x20, 0x32, 0x30, 0x31, 0x30, 0x20, 0x41, 0x63, 0x74, 0x69, 0x76, 0x69, 0x73, 0x69,
+      0x6F, 0x6E, 0x2E, 0x20, 0x41, 0x6C, 0x6C, 0x20, 0x52, 0x69, 0x67, 0x68, 0x74, 0x73,
+      0x20, 0x52, 0x65, 0x73, 0x65, 0x72, 0x76, 0x65, 0x64, 0x2E, 0x20};
+
+  memcpy(hash_in.data() + 0x21, HASH_CONST.data(), HASH_CONST.size());
+
+  memcpy(dest, hash_in.data(), 0x56);
+}
+void SkylanderFigure::Encrypt(std::span<u8, FIGURE_SIZE> input)
+{
+  std::array<u8, 0x56> hash_in = {};
+
+  GenerateIncompleteHashIn(hash_in.data());
+
+  std::array<u8, FIGURE_SIZE> encrypted = {};
+
+  std::array<u8, 0x10> current_block = {};
+
+  // Run for every block
+  for (u8 i = 0; i < 64; ++i)
+  {
+    memcpy(current_block.data(), input.data() + (i * 0x10), 0x10);
+
+    // Skip sector trailer and the first 8 blocks
+    if (((i + 1) % 4 == 0) || i < 8)
+    {
+      memcpy(encrypted.data() + (i * 0x10), current_block.data(), 0x10);
+      continue;
+    }
+
+    // Block index
+    hash_in[0x20] = i;
+
+    std::array<u8, 0x10> hash_out = {};
+
+    mbedtls_md5_ret(hash_in.data(), 0x56, hash_out.data());
+
+    mbedtls_aes_context aes_context = {};
+
+    mbedtls_aes_setkey_enc(&aes_context, hash_out.data(), 128);
+
+    mbedtls_aes_crypt_ecb(&aes_context, MBEDTLS_AES_ENCRYPT, current_block.data(),
+                          encrypted.data() + (i * 0x10));
+  }
+
+  memcpy(m_data.data(), encrypted.data(), FIGURE_SIZE);
+
+  NOTICE_LOG_FMT(IOS_USB, "Encrypted skylander data: \n{}", HexDump(encrypted.data(), FIGURE_SIZE));
+}
+SkylanderFigure::SkylanderFigure(File::IOFile file)
+{
+  m_sky_file = std::move(file);
+  m_sky_file.Seek(0, File::SeekOrigin::Begin);
+  m_sky_file.ReadBytes(m_data.data(), m_data.size());
+}
+bool SkylanderFigure::Create(u16 m_sky_id, u16 m_sky_var)
+{
+  if (!m_sky_file)
+  {
+    return false;
+  }
+
+  memset(m_data.data(), 0, m_data.size());
+
+  PopulateSectorTrailers();
+
+  // Set the NUID of the figure
+  Common::Random::Generate(&m_data[0], 4);
+
+  // The BCC (Block Check Character)
+  m_data[4] = m_data[0] ^ m_data[1] ^ m_data[2] ^ m_data[3];
+
+  // ATQA
+  m_data[5] = 0x81;
+  m_data[6] = 0x01;
+
+  // SAK
+  m_data[7] = 0x0F;
+
+  // Set the skylander info
+  memcpy(&m_data[0x10], &m_sky_id, sizeof(m_sky_id));
+  memcpy(&m_data[0x1C], &m_sky_var, sizeof(m_sky_var));
+
+  // Set checksum
+  ComputeChecksum(ChecksumType::Type0, m_data.data(), m_data.data() + 0x1E);
+
+  PopulateKeys();
+
+  Save();
+  return true;
+}
+void SkylanderFigure::Save()
+{
+  m_sky_file.Seek(0, File::SeekOrigin::Begin);
+  m_sky_file.WriteBytes(m_data.data(), 0x40 * 0x10);
+}
+
+void SkylanderFigure::GetBlock(u8 index, u8* dest) const
+{
+  memcpy(dest, m_data.data() + (index * 0x10), 0x10);
+}
+
+FigureData SkylanderFigure::GetData() const
```

Comment:
This function is *full* of `reinterpret_cast`s that I believe could use `Common::BitCast` instead, that way the behavior is actually defined correctly.

Reply:
OH. I was already wondering why that was better.